tensorflow/compiler/jit/xla_compilation_cache.cc

```diff
@@ -317,12 +321,112 @@ void LogOnceXlaCompiledFirstCluster() {
+
+  entry->compilation_status =
+      compile_fn(&compiler, args, &entry->compilation_result);
+  TF_RETURN_IF_ERROR(entry->compilation_status);
+  CHECK_EQ(entry->executable.get(), nullptr);
+  entry->compilation_status =
+      BuildExecutable(options, entry->compilation_result, &entry->executable);
+
+  const uint64 compile_end_us = env->NowMicros();
+  const uint64 compile_time_us = compile_end_us - compile_start_us;
+  metrics::UpdateXlaCompilationTime(compile_time_us);
+  {
+    mutex_lock lock(cluster_compile_stats_mu_);
+    auto it = cluster_compile_stats_.find(function_name);
+    const uint64 compile_time_s = compile_time_us / 1.0e6;
+    it->second.compile_count++;
+    it->second.cumulative_compile_time_us += compile_time_us;
+    it->second.max_compile_time_s = std::max(it->second.max_compile_time_s,
+                                             compile_time_s);
+    LogOnceXlaCompiledFirstCluster();
+    VLOG(1) << "compiled " << function_name << " "
+            << it->second.compile_count
+            << " times, compile time: " << compile_time_us
+            << " us, cumulative: " << it->second.cumulative_compile_time_us
+            << " us ("
+            << tensorflow::strings::HumanReadableElapsedTime(compile_time_s)
+            << " / "
+            << tensorflow::strings::HumanReadableElapsedTime(
+                   it->second.cumulative_compile_time_us / 1.0e6)
+            << ")";
+
+    XlaJitCompilationActivity jit_compilation_activity;
+    jit_compilation_activity.set_cluster_name(function_name);
+    jit_compilation_activity.set_compile_count(it->second.compile_count);
+    jit_compilation_activity.set_compile_time_us(compile_time_us);
+    jit_compilation_activity.set_cumulative_compile_time_us(
+        it->second.cumulative_compile_time_us);
+    TF_RETURN_IF_ERROR(
+        BroadcastXlaActivity(std::move(jit_compilation_activity)));
+  }
+
+  return Status::OK();
+}
+
+Status XlaCompilationCache::CompileAsynchronous(
+    Entry* entry, const XlaCompiler::Options& options,
+    const std::vector<XlaCompiler::Argument>& args,
+    const string &function_name,
+    const std::function<Status(XlaCompiler* compiler,
+                               const std::vector<XlaCompiler::Argument>& args,
+                               XlaCompiler::CompilationResult*)>& compile_fn) {
+  entry->compile_state = CompileState::kCompiling; // Still under caller's lock.
+  {
+    mutex_lock lock(async_compilation_.async_compilation_mu_);
+    async_compilation_.num_ongoing_compilations++;
+  }
+  // Don't move the above code into the thread function!!!
+
+  // Passing options by value into the lamba increases the refcount on
+  // options.device_allocator, keeping it alive for the duration of the
+  // compilation.
+  // Passing args by value as well. Doing this here only when an asynchronous
+  // compilation is performed, as copying many args incurs an overhead.
+  async_compilation_.compiler_threads.Schedule([=] {
```

Comment:
Also, are we assuming that `Schedule` always uses a different thread (since otherwise `mutex_lock entry_lock(entry->mu);` would deadlock)?

Another question: can we unify the strict and async compilation by making `Schedule` immediately execute the closure in the former case?  Or would the overhead of copying `args` be too much?

Reply:
Schedule uses its own threadpool, that cannot conflict with the thread used by CompileImpl